Process KeyUpdate and NewSessionTicket messages after a close_notify

mattcaswell opened the pull request:
If we've sent a close_notify then we are restricted about what we can do in response to handshake messages that we receive. However we can sensibly process NewSessionTicket messages. We can also process a KeyUpdate message as long as we also ignore any request for us to update our sending keys.

This is an alternative to #7058. I've used the same test proposed in that PR, so @kroeckx is shown as the author in the second commit.

mattcaswell commented:
Ping? This is on the critical path for the 1.1.1 release.

kroeckx commented on test/sslapitest.c:
```diff
@@ -5400,13 +5405,22 @@ static int test_shutdown(int tst)
+                || !TEST_true(SSL_write(serverssl, msg, sizeof(msg))))
+            goto end;
+        if (tst == 4 &&
+                (!TEST_true(SSL_key_update(serverssl, SSL_KEY_UPDATE_NOT_REQUESTED))
```
Can you change that to requested?

mattcaswell replied:
Good idea. Will do.

kroeckx commented on ssl/record/rec_layer_s3.c:
```diff
@@ -1554,30 +1554,44 @@ int ssl3_read_bytes(SSL *s, int type, int *recvd_type, unsigned char *buf,
+                    || (mt != SSL3_MT_NEWSESSION_TICKET
+                        && mt != SSL3_MT_KEY_UPDATE)) {
+                SSL3_RECORD_set_length(rr, 0);
+                SSL3_RECORD_set_read(rr);
```
I would argue that instead of just throwing it away, we should go to the error state. An application should not bring us in this state ever.

kroeckx replied:
In the other PR you quoted this about client authentication: `between sending the CertificateRequest and receiving a response` The client should still get authenticated, the server is not going to request client authentication if it's happy to just not authenticate. If the client already send a shutdown, and then is being asked to authenticate, the server is going to fail, I see no point in the client not also failing.

mattcaswell replied:
> An application should not bring us in this state ever.

Yes it can. As I previously mentioned it is perfectly possible for the peer to have not read our close_notify yet even though we have sent it.

> The client should still get authenticated, the server is not going to request client authentication if it's happy to just not authenticate.

Refusal to authenticate is not necessary fatal to an application. It may just mean that access to certain things will not be possible. In fact the RFC allows clients to refuse to authenticate.

> If the client already send a shutdown, and then is being asked to authenticate, the server is going to fail, I see no point in the client not also failing.

Those are application level decisions which are not up to us to decide.

kroeckx commented on ssl/record/rec_layer_s3.c:
```diff
@@ -1554,30 +1554,44 @@ int ssl3_read_bytes(SSL *s, int type, int *recvd_type, unsigned char *buf,
+             */
+            if (!SSL_IS_TLS13(s)
+                       /* Shouldn't ever be 0 but lets be safe */
+                    || SSL3_RECORD_get_length(rr) == 0
```
In case this == 0, isn't the get_data() above wrong?

kaduk replied:
Yes, but Matt claims that this shouldn't ever be zero.  My intuition agrees with Matt, though I am not finding where in the code this is presently enforced.

kroeckx commented on ssl/record/rec_layer_s3.c:
```diff
@@ -1554,30 +1554,44 @@ int ssl3_read_bytes(SSL *s, int type, int *recvd_type, unsigned char *buf,
+             */
+            unsigned char mt = s->rlayer.handshake_fragment_len > 0
+                               ? s->rlayer.handshake_fragment[0]
+                               : *SSL3_RECORD_get_data(rr);
```
I think handshake_fragment_len == 0 if type == SSL3_RT_HANDSHAKE, according to a comment above in the function.

kroeckx replied:
A later comment then says: s->rlayer.handshake_fragment_len == 4  iff  rr->type == SSL3_RT_HANDSHAKE

Clearly I don't understand this code good enough.

kaduk replied:
```type``` is the type of data that the caller is requesting; in the cases considered here it will be application data.
(The actual record we're processing would be a handshake message, though, of course.)

I'm not 100% conversant with this code, either, but I do not really believe that the comment you quote is accurate at the location it is currently placed, since we populate handshake_fragment later in the function at present.

mattcaswell replied:
As @kaduk says the comment you refer to only applies if `type == SSL3_RT_HANDSHAKE` which is only true if the caller is asking for handshake data - which isn't the case here.

> I'm not 100% conversant with this code, either, but I do not really believe that the comment you quote is accurate at the location it is currently placed, since we populate handshake_fragment later in the function at present.

The function loops. If we receive handshake data that is less than `SSL3_HM_HEADER_LENGTH` in length then we go back to `start` (see the comment "fragment was too small"). It then attempts to get another record. That could of course end up returning to the calling application if no record is available. At that point the application could send `close_notify`. If another read then occurs and more handshake data has arrived then we will get to this point in the code and `handshake_fragment_len` will be > 0.

kroeckx commented:
I think we disagree on how to fix this. I also feel like I don't understand the code good enough to be able to review it. I suggest you find somebody else to review it.

kaduk commented on ssl/record/rec_layer_s3.c:
```diff
@@ -1586,10 +1600,12 @@ int ssl3_read_bytes(SSL *s, int type, int *recvd_type, unsigned char *buf,
                      SSL_R_APPLICATION_DATA_AFTER_CLOSE_NOTIFY);
+            return -1;
         }
-        return -1;
```
It's unclear why we need to pull the ```return -1``` into both arms of the conditional.

mattcaswell replied:
Because in the first arm there is a codepath that does not return -1 and we continue. In fact that is the whole point. If we receive KeyUpdate or NewSessionTicket we want to continue onwards to process those messages without returning with -1.

mattcaswell commented:
> I think we disagree on how to fix this.

Do we? I haven't heard an objection to the approach - only debate on some of the details. But ok.

> I also feel like I don't understand the code good enough to be able to review it. I suggest you find somebody else to review it.

Ok.

mattcaswell commented:
New commits pushed addressing the feedback so far. On looking at the whole message fragment piece again I realised that it was a bit of a can of worms and there were some corner cases I hadn't considered. Therefore I removed that piece and simplified it so that we simply ignore all handshake messages for <= TLSv1.2 and process all TLSv1.3 handshake messages. Since we're talking post-handshake only, that only actually means that we additionally need to handle PHA CertificateRequest in the state machine (i.e. to ignore it if we already sent close_notify).

Ready for more review.

mattcaswell commented:
Pushed. Thanks.

mspncp commented:
> Pushed. Thanks.

Are you sure? I don't see any pushed commit.

mattcaswell commented:
Well done for paying attention. You passed my test. Pushed for real this time. ;-)

mspncp commented:
> Well done for paying attention. 

Thanks! It's a habit of mine to wait for the "levitte pushed a commit..." to appear on the github page before closing the pr.
